HELP.md

```diff
@@ -0,0 +1,761 @@
+#[near_bindgen]
+impl Contract {
+    #[private]
+    pub fn resolve_transfer(&mut self) {
+        env::log(b"This is a callback");
+    }
+}
+```
+
+This is equivalent to:
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn resolve_transfer(&mut self) {
+        if env::current_account_id() != env::predecessor_account_id() {
+            near_sdk::env::panic(b"Method resolve_transfer is private");
+        }
+        env::log(b"This is a callback");
+    }
+}
+```
+
+## Integer JSON types
+
+NEAR Protocol currently expects contracts to support JSON serialization. JSON can't handle large integers (above 2**53 bits).
+That's why you should use helper classes from the `json_types` in `near_sdk` for `u64` and `u128`.
+We provide types `U64` and `U128`, that wraps the integer into a struct and implements JSON serialization and
+deserialization as a base-10 strings.
+
+You can convert from `U64` to `u64` and back using `std::convert::Into`, e.g.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn mult(&self, a: U64, b: U64) -> U128 {
+        let a: u64 = a.into();
+        let b: u64 = b.into();
+        let product = u128::from(a) * u128::from(b);
+        product.into()
+    }
+}
+```
+
+You can also access inner values and using `.0` and `U128(5)`, e.g.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn sum(&self, a: U128, b: U128) -> U128 {
+        U128(a.0 + b.0)
+    }
+}
+```
+
+## `Base64VecU8` JSON type
+
+Often the contract needs to receive or return a binary data.
+Encoding a `Vec<u8>` with JSON will lead to an integer array, e.g. `[110, 101, 97, 114]`
+This is inefficient in both compute and space.
+
+`Base64VecU8` is a wrapper on top of `Vec<u8>` that allows to pass it as arguments or result.
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    // Notice, internally we store `Vec<u8>` 
+    pub data: Vec<u8>,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(data: Base64VecU8) -> Self {
+        Self {
+            data: data.into(),
+        }
+    }
+
+    pub fn get_data(self) -> Base64VecU8 {
+        self.data.into()
+    }
+}
+```
+
+## View vs Change method
+
+`near_sdk` assumes that the method is a `view` if it uses `&self` or `self` and method is `change` if it has `&mut self`.
+
+View methods don't save the contract STATE at the end of the method execution.
+
+Change methods will save the modified STATE at the end of the method execution. They can also modify the state in persistent collections.
+
+Note: Change methods will also check that the function call doesn't have attached deposit, unless the method is marked with the `#[payable]` macro.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    /// View method. Requires cloning the account id.
+    pub fn get_owner_id(&self) -> AccountId {
+        self.owner_id.clone()
+    }
+
+    /// View method. More efficient, but can't be reused internally, because it consumes self.
```

Comment:
Is there a reason to prefer Rust's doc comments with three slashes to non-doc comments with only two?

Reply:
There's a good example here, and I think this is typically helpful as it is here right above a function. A non-doc comment might be in the implementation if something seems unintuitive and it's helpful to add a note.
https://doc.rust-lang.org/reference/comments.html#doc-comments

Reply:
Ok, but the docs are not really _used_ yet, in the NEAR ecosystem. 

I'm guessing that someday the Metadata API might use doc comments.

Reply:
I don't think that's accurate, we just haven't fully fleshed it out yet, which I hope we'll do this quarter.
<img width="2089" alt="Screen Shot 2021-04-09 at 2 02 01 PM" src="https://user-images.githubusercontent.com/1042667/114240374-43d43400-993c-11eb-80d1-a33a7c9e7b31.png">